Reveal.js 3.8.0 to 3.9.2 support

obilodeau opened the pull request:
Draft commit for reveal.js 3.8.0 support.

This work is postponed to have a very stable 3.0.0 that will support reveal.js versions spanning from 3.0 to 3.7. See #296.

Needs work:
- [x] syntax highlight `<mark>` syntax [described here](https://github.com/hakimel/reveal.js/blob/master/README.md#code-syntax-highlighting) does not work
- [x] line number highlights [described here](https://github.com/hakimel/reveal.js/blob/master/README.md#line-numbers--highlights) do not work
- [x] default plugin call for highlightjs is not current but it doesn't seem to break anything
- [ ] ~I'm not sure per-slide iframe preloading options are supported~
- [x] Also, see commits messages and introduced comments for more to add to the task list
- [x] test `hash` feature w/ and w/o history and see if direct links still works

obilodeau commented:
* Rebased on top of 3.1.0
* Aligned with latest reveal.js `index.html` template for slide decks (dropped IE hacks, safari viewport stuff, etc.)
* Highlightjs changes (the way its called, default css to align with upstream)
* Dropped redundant print-pdf plugin call

Also, highlightjs `<mark>` does work if we put valid html angle brackets in the generated source file so we need to tell Asciidoctor to let them through. Result is ugly though but it seems to be intended (I tested with only reveal.js).

obilodeau commented:
Tonight's work:
* highlightjs `<mark>` works with `subs="none"`, adjusted the example accordingly
* highlightjs line highlight works, it kicks-in line highlighting which might be unexpected and there seems to be a CSS activated that breaks the output. Still need to work on this.

![highlightjs-highlight-ugly](https://user-images.githubusercontent.com/546325/73420621-23005f80-42f1-11ea-87f3-661400e5d3fc.png)

* added broken examples for other supported highlighters, given I have little time to spend on this project (and I'm spending a lot right now) I won't focus on them but they are documented as broken and people can step in

obilodeau commented on lib/asciidoctor-revealjs/highlightjs.rb:
```diff
@@ -11,10 +11,20 @@ def initialize *args
+            # Note for review: should the API be modified to give easier access to this?
+            if node.attributes.key?("highlight")
+              code['data-line-numbers'] = self._convert_linedef_to_revealjs(node.attributes["highlight"])
+            end
```
I don't think my access to the `highlight` attribute of the listing block is proper here. It should go through the syntax highlight API but I didn't find a way to do so. `opts[:highlight_lines]` is available in `highlight()` but not in `format()`.

@Mogztter, @mojavelinux thoughts?

Mogztter replied:
Sorry I can't help, I'm not familiar with this API so I don't know the reason behind this choice

mojavelinux replied:
I didn't think about the fact that the option would be needed in format. We could take that up in core. In the meantime, you can call:

```
resolve_lines_to_highlight source, (attr 'highlight')
```

mojavelinux replied:
don't do it this way:

```
highlight_lines.gsub("..", "-")
```

The logic is much more complex than that. See https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/substitutors.rb#L968-L999

obilodeau replied:
The problem is that reveal.js introduced a new legal character `|` to split highlights into steps sequences. It's a very cool feature that I intend to use in presentations. Documented here: https://github.com/hakimel/reveal.js#step-by-step-highlights

So I'm going to need to split on it, then send each section to `resolve_lines_to_highlight` method and rejoin everything after. I'm on it right now.

In the meantime, here how I adapted your code to make it work, in case you have any comment:

```ruby
if node.attributes.key?("highlight")
    code['data-line-numbers'] = node.resolve_lines_to_highlight(node.content, node.attributes["highlight"]).join(',')
```

obilodeau replied:
Fixed in a839f53. This is probably the most idiomatic ruby I've written so far. Marking this as resolved but allowing you to further comment for another day before merging.

Mogztter commented on examples/source-rouge.adoc:
```diff
@@ -39,3 +40,20 @@ fn main() {
 }
 ----
+
+== Rouge Reatures
```
```suggestion
== Rouge Features
```

Mogztter commented on examples/source-rouge.adoc:
```diff
@@ -39,3 +40,20 @@ fn main() {
+// TODO docinfo not loaded
+// TODO linenums feature affects listing size
+
+This is broken
```
Actually, it's a limitation of Rouge: https://github.com/rouge-ruby/rouge/issues/264
I think we shouldn't bother too much.

obilodeau replied:
I'm not going to do it.

Mogztter commented on examples/source-rouge.adoc:
```diff
@@ -39,3 +40,20 @@ fn main() {
+
+== Rouge Reatures
+
+// TODO docinfo not loaded
```
This sounds more concerning...

obilodeau replied:
docinfo is there now thanks to the rebase on top of #324. However, the class `hll` is missing from the styles inlined by Asciidoctor. Again, anything else than highlight.js is not getting my attention right now.

Mogztter commented:
@obilodeau How can I help with this pull request? If you have specific tasks, please tell me. Hopefully I've fixed the docinfo issue in #324

obilodeau commented:
Rebased on top of #324.

obilodeau commented:
Highlight.js is feature parity with 3.8.0. I'm going to deal with 3.8.0 to ~3.9.1~ 3.9.2 later.

Next-up: go through all configuration options and see if I'm missing anything

obilodeau commented:
* Added new configuration options
* Added `previewLinks` and `data-preview-link` support
* Rebased on top of #325
* Started upgrade notes and added a compatibility matrix to the README

obilodeau commented:
Pushed doctests

obilodeau commented:
Lazy loading is not supported at the moment and would require a lot of testing and changes. I have video-heavy presentations and never felt it was required. Also, it can be implemented later without causing a breaking change. Let's not implement it now.

obilodeau commented:
This is ready to go in. I have a big presentation coming up so I would like to release this relatively shortly (in about a week max).

@Mogztter if you want #320 to get in for 4.0.0, I would advise you to get it in shape quickly. I would prefer it if it does get in. Plus there is a lot of doctest around highlightjs so I feel confident.

@mojavelinux there's an API question that you might want to look at in this review

Lastly, @Mogztter since you asked how you can help, there's #326 that I would like to get done soon. I'm not very good at maintainable CSS so if you can give it a try I would appreciate it.

Mogztter commented:
Speaking of #326, I need to do some vertical alignment and as far as I know the only way to do it reliably is to wrap the content of a slide in a `<div>`. Is that something you would consider or should I override the "section" template and do it in my presentation?

obilodeau commented:
I had the feeling that it was achievable without doing so. Isn't it working
for the other example I provided in the issue?

Mogztter commented:
I don't want to pollute the pull request about the reveal.js upgrade, so here's my reply: https://github.com/asciidoctor/asciidoctor-reveal.js/issues/326#issuecomment-582480509

Mogztter commented on templates/document.html.slim:
```diff
@@ -53,15 +55,15 @@ html lang=(attr :lang, 'en' unless attr? :nolang)
       =syntax_hl.docinfo :head, self, cdn_base_url: cdn_base, linkcss: linkcss, self_closing_tag_slash: '/'
-    / If the query includes 'print-pdf', use the PDF print sheet
+
+    /! Printing and PDF exports
```
Do we really want an HTML comment here?

obilodeau replied:
I'm exactly following reveal.js' presentation frame

Mogztter replied:
All good then 👍

Mogztter commented on test/doctest/theme-custom.html:
```diff
@@ -133,39 +165,37 @@ <h2>Actually, they are…​</h2>
     // The display mode that will be used to show slides
     display: 'block',
-
+  
```
can we remove the trailing spaces?

obilodeau replied:
Sure, `rake doctest:generate` creates them and I was too lazy to remove them. If `doctest` doesn't mind (and I my guess is it doesn't) then I'll remove them.

obilodeau replied:
Fixed in e5b8b64

Mogztter commented:
> if you want #320 to get in for 4.0.0, I would advise you to get it in shape quickly. I would prefer it if it does get in. Plus there is a lot of doctest around highlightjs so I feel confident.

Alright, I will work on it tomorrow 😉

obilodeau commented:
Fixed whitespace
